### PR TITLE
HTML API: Fix detection of parsing namespace in integration nodes

### PR DESCRIPTION
**Pull request by sirreal**
Trac ticket: Core-61576

Fix an issue where integration points do not correctly adjust the Tag Processor's parsing namespace on enter/exit.

When the adjusted current node changes, the Tag Processor's parsing namespace should be adjusted accordingly. This ensures that rules like an `<image>` tag in HTML is transformed into ``.

This can be observed with the `image` tags in this example HTML:

```html
<svg><foreignobject><image/><svg/><image/></svg>
<math><annotation-xml encoding=text/html><image/><math/><image/>
```

All of the `image` tags in that html should be handled following the html parsing rules as `img`. This currently fails on trunk.

[Compare this branch](https://playground.wordpress.net/?plugin=html-api-debugger&php-extension-bundle=light&url=%2Fwp-admin%2Fadmin.php%3Fpage%3Dhtml-api-debugger%26html%3D%253Csvg%253E%253Cforeignobject%253E%253Cimage%252F%253E%253Csvg%252F%253E%253Cimage%252F%253E%253C%252Fsvg%253E%250A%253Cmath%253E%253Cannotation-xml%2Bencoding%253Dtext%252Fhtml%253E%253Cimage%252F%253E%253Cmath%252F%253E%253Cimage%252F%253E&wp=https%3A%2F%2Fplayground.wordpress.net%2Fplugin-proxy.php%3Forg%3DWordPress%26repo%3Dwordpress-develop%26workflow%3DTest%2BBuild%2BProcesses%26artifact%3Dwordpress-build-7425%26pr%3D7425) with [nightly](https://playground.wordpress.net/?plugin=html-api-debugger&php-extension-bundle=light&url=%2Fwp-admin%2Fadmin.php%3Fpage%3Dhtml-api-debugger%26html%3D%253Csvg%253E%253Cforeignobject%253E%253Cimage%252F%253E%253Csvg%252F%253E%253Cimage%252F%253E%253C%252Fsvg%253E%250A%253Cmath%253E%253Cannotation-xml%2Bencoding%253Dtext%252Fhtml%253E%253Cimage%252F%253E%253Cmath%252F%253E%253Cimage%252F%253E&wp=nightly).

```diff
- Tests: 2490, Assertions: 4184, Skipped: 422.
+ Tests: 2492, Assertions: 4203, Skipped: 422.
```

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/html-api/class-wp-html-processor.php [hunk 1]
```diff
@@ -403,12 +403,13 @@ function ( WP_HTML_Token $token ): void {
...
-				$this->change_parsing_namespace(
-					$adjusted_current_node
-						? $adjusted_current_node->namespace
-						: 'html'
-				);
+
+				if ( $adjusted_current_node ) {
+					$this->change_parsing_namespace( $adjusted_current_node->integration_node_type ?? $adjusted_current_node->namespace );
+				}
+				$this->change_parsing_namespace( 'html' );
```
Comment (dmsnell) [inline src/wp-includes/html-api/class-wp-html-processor.php]: @sirreal was this line a mistake? we're _always_ setting the parsing namespace to HTML…

Comment (sirreal) [reply src/wp-includes/html-api/class-wp-html-processor.php]: Yes! That certainly appears to be a mistake, I think this was intended to be `if`/`else`.

**Comment by sirreal** [comment]
**Edit: removed**

<details>
<summary>redacted</summary>

@dmsnell My first thought was that this logic should be used to determine the parsing mode before calling Tag_Processor (parent) `step`:

https://github.com/WordPress/wordpress-develop/blob/34c861b6a53cb82721ce0e82a3e2024d91c6ef7e/src/wp-includes/html-api/class-wp-html-processor.php#L873-L892

And _maybe_ the tag procuessor just needs to know `html`/`foreign` more than the actual namespace it's in 🤔 

</details>

File: src/wp-includes/html-api/class-wp-html-processor.php [hunk 2]
```diff
@@ -393,7 +393,7 @@ function ( WP_HTML_Token $token ): void {
 				$provenance            = ( ! $same_node || $is_virtual ) ? 'virtual' : 'real';
 				$this->element_queue[] = new WP_HTML_Stack_Event( $token, WP_HTML_Stack_Event::PUSH, $provenance );
 
-				$this->change_parsing_namespace( $token->namespace );
+				$this->change_parsing_namespace( $token->integration_node_type ? 'html' : $token->namespace );
```
Comment (sirreal) [inline src/wp-includes/html-api/class-wp-html-processor.php:396]: Good spot, I think you're correct on this. I hadn't appreciated how integration_node_type works 👍

**Comment by sirreal** [comment]
Thanks for the fixes here, I think this is good to go @dmsnell.

**Comment by dmsnell** [comment]
Merged in [[59099]](https://core.trac.wordpress.org/changeset/59099)
https://github.com/wordpress/wordpress-develop/commit/1e21ecedf19f1b97360949a9e509a3c04ac1f34e

**Pull request closed without merging**